tests/queries/0_stateless/01515_with_global_and_with_propagation.sql

```diff
@@ -1,13 +1,15 @@
 SET enable_global_with_statement = 1;
 
-WITH 1 AS x SELECT x;
-WITH 1 AS x SELECT * FROM (SELECT x);
-WITH 1 AS x SELECT *, x FROM (WITH 2 AS x SELECT x AS y);
-WITH 1 AS x SELECT x UNION ALL SELECT x;
-select x from (WITH 1 AS x SELECT x UNION ALL WITH 2 AS x SELECT x) order by x;
+with 1 as x select x;
+with 1 as x select * from (select x);
+with 1 as x select *, x from (with 2 as x select x as y);
+with 1 as x select x union all select x;
+select x from (with 1 as x select x union all with 2 as x select x) order by x;
+with 5 as q1, x as (select number+100 as b, number as a from numbers(10) where number > q1) select * from x;
 
-EXPLAIN SYNTAX WITH 1 AS x SELECT x;
-EXPLAIN SYNTAX WITH 1 AS x SELECT * FROM (SELECT x);
-EXPLAIN SYNTAX WITH 1 AS x SELECT *, x FROM (WITH 2 AS x SELECT x AS y);
-EXPLAIN SYNTAX WITH 1 AS x SELECT x UNION ALL SELECT x;
-EXPLAIN SYNTAX WITH 1 AS x SELECT x UNION ALL WITH 2 AS x SELECT x;
+explain syntax with 1 as x select x;
+explain syntax with 1 as x select * from (select x);
+explain syntax with 1 as x select *, x from (with 2 as x select x as y);
+explain syntax with 1 as x select x union all select x;
+explain syntax with 1 as x select x union all with 2 as x select x;
+explain syntax with 5 as q1, x as (select number + 100 as b, number as a from numbers(10) where number > q1) select * from x;
```

Comment:
Right now the following query will lead to endless recursion and eventually fails due to memory limits:

```sql
SELECT *
FROM 
(
    WITH
        
        (
            SELECT 1
        ) AS r1,
        
        (
            SELECT 2
        ) AS r2,
        
        (
            SELECT 3
        ) AS r3
    SELECT 10
)
SETTINGS max_ast_elements = 1000, max_memory_usage = '10Mi'
```

And this is because of `ExecuteScalarSubqueriesMatcher`, that visits AST and create `InterpreterSelectQuery` for each subquery (`ASTSubquery`), while `ApplyWithAliasVisitor` will create more each time.

So how about the following patch?

```patch
$ git di src/Interpreters/ExecuteScalarSubqueriesVisitor.cpp
diff --git a/src/Interpreters/ExecuteScalarSubqueriesVisitor.cpp b/src/Interpreters/ExecuteScalarSubqueriesVisitor.cpp
index b3e8af90f2..4a6ec3505c 100644
--- a/src/Interpreters/ExecuteScalarSubqueriesVisitor.cpp
+++ b/src/Interpreters/ExecuteScalarSubqueriesVisitor.cpp
@@ -92,6 +92,12 @@ void ExecuteScalarSubqueriesMatcher::visit(const ASTSubquery & subquery, ASTPtr
         Settings subquery_settings = data.context.getSettings();
         subquery_settings.max_result_rows = 1;
         subquery_settings.extremes = false;
+        /// Otherwise ApplyWithAliasVisitor will be applied and this will
+        /// create new nodes in AST over and over and lead to endless recursion,
+        /// which eventually lead to one of the following errors:
+        /// - Memory limit exceeded
+        /// - AST too big
+        subquery_settings.enable_global_with_statement = false;
         subquery_context.setSettings(subquery_settings);
 
         ASTPtr subquery_select = subquery.children.at(0);
```

Also not that there is also `DB::JoinedTables::resolveTables` that will create `InterpreterSelectQuery`, and so it will also expand aliases one more time, which is pointless.

So maybe it is worth disabling `enable_global_with_statement` once `ApplyWithAliasVisitor` was applied (with changed=false)? Since there are also other places where interpreter will be created multiple times.

Reply:
Yeah, I'm fixing this and other related issues. Thanks for the suggestion. Actually we can just disable with propagation when `subquery_depth` is not zero.